### ISSUE TITLE
Testing size-based rotation, doesn't seem to work as expected

### ISSUE THREAD

**Issue by ErisDS** [log4js-node/log4js-node#911]
I'm still running tests on log4js assessing it's viability as a replacement for the no longer maintained bunyan.

I just did a very very basic test on size-based rotation, and I'm not seeing the behaviour I'd expect.

Here's my test code:

```
const log4js = require('log4js');
const config = {
    appenders: {
        access: {type: 'file', filename: 'logs/test.log', maxLogSize: 5000, backups: 4}
    },
    categories: {default: {appenders: ['access'], level: 'debug'}}
};

log4js.configure(config);

const logger = log4js.getLogger('default');

logger.info('hello world 1');
logger.error(new Error('bad thing 1'));
logger.info('hello world 2');
logger.error(new Error('bad thing 2'));
```

This file is saved as `logging.js` and I'm hammering `node logging.js` locally on my macbook pro. I'm expecting there to quickly be 10 files, but there are never more than 2. 

If I duplicate my log lines at the bottom so much more log content is output, I can get 10 files to be created, but never from consistently hitting the script.

I thought it might be related to this bug, because outputting more content in one stream/session creates more files: https://github.com/log4js-node/streamroller/issues/34 - but it seems to be entirely to do with the number of backup files created, not the size of the individual files.

I'm looking for file size based rotation, not date based rotation, due the sheer volume of log files I'm going to be using this with - rotating all the log files across 1000s of apps at midnight causes performance problems 😬

**Comment by ErisDS** [comment]
Here's the DEBUG output from the point at which I'd expect test.log.2 to appear:

```
 streamroller:RollingFileWriteStream creating RollingFileWriteStream. path=logs/test.log +0ms
  streamroller:RollingFileWriteStream creating stream with option={"type":"file","filename":"logs/test.log","maxLogSize":5000,"backups":4,"maxSize":5000,"numToKeep":4,"encoding":"utf8","mode":420,"flags":"a","compress":false,"keepFileExt":false,"alwaysIncludePattern":false} +1ms
  streamroller:RollingFileWriteStream _formatFileName: date=undefined, index=undefined, isHotFile=true +0ms
  streamroller:RollingFileWriteStream _formatFileName: includePattern? false, pattern: undefined +3ms
  streamroller:RollingFileWriteStream _formatFileName: is hot file so, filename: test.log +0ms
  streamroller:RollingFileWriteStream create new file with no hot file. name=test.log, state={"currentDate":"2019-07-04T18:39:33.490Z","currentIndex":0,"currentSize":5086} +1ms
  streamroller:RollingFileWriteStream _formatFileName: date=Thu Jul 04 2019 19:39:33 GMT+0100 (British Summer Time), index=undefined, isHotFile=true +1ms
  streamroller:RollingFileWriteStream _formatFileName: includePattern? false, pattern: undefined +0ms
  streamroller:RollingFileWriteStream _formatFileName: is hot file so, filename: test.log +0ms
  streamroller:RollingFileWriteStream in _shouldRoll, pattern = undefined, currentDate = Thu Jul 04 2019 19:39:33 GMT+0100 (British Summer Time), now = Thu Jul 04 2019 19:39:40 GMT+0100 (British Summer Time) +4ms
  streamroller:RollingFileWriteStream in _shouldRoll, currentSize = 5086, maxSize = 5000 +0ms
  streamroller:RollingFileWriteStream rolling, isNextPeriod ? false +0ms
  streamroller:RollingFileWriteStream _roll: closing the current stream +0ms
  streamroller:RollingFileWriteStream currentIndex = 0, numToKeep = 4 +3ms
  streamroller:RollingFileWriteStream i = 0 +0ms
  streamroller:RollingFileWriteStream _formatFileName: date=Thu Jul 04 2019 19:39:33 GMT+0100 (British Summer Time), index=1, isHotFile=undefined +0ms
  streamroller:RollingFileWriteStream _formatFileName: test.log.1 +0ms
  streamroller:RollingFileWriteStream src=logs/test.log, tgt=logs/test.log, idx=0, pos=0 +1ms
  streamroller:RollingFileWriteStream moveAndMaybeCompressFile: moving file from logs/test.log to logs/test.log.1 without compress +0ms
  streamroller:RollingFileWriteStream moveAndMaybeCompressFile: deleting file=logs/test.log.1, renaming logs/test.log to logs/test.log.1 +0ms
  streamroller:RollingFileWriteStream rolling during the same period. state={"currentDate":"2019-07-04T18:39:33.490Z","currentIndex":1,"currentSize":0} +2ms
  streamroller:RollingFileWriteStream _formatFileName: date=Thu Jul 04 2019 19:39:33 GMT+0100 (British Summer Time), index=undefined, isHotFile=true +0ms
  streamroller:RollingFileWriteStream _formatFileName: includePattern? false, pattern: undefined +0ms
  streamroller:RollingFileWriteStream _formatFileName: is hot file so, filename: test.log +0ms
  streamroller:RollingFileWriteStream _getExistingFiles: files=test.log,test.log.1 +1ms
  streamroller:RollingFileWriteStream metaStr=, fileName=test.log, prefix=test.log +1ms
  streamroller:RollingFileWriteStream _getExistingFiles: parsed test.log as [object Object] +0ms
  streamroller:RollingFileWriteStream metaStr=.1, fileName=test.log.1, prefix=test.log +0ms
  streamroller:RollingFileWriteStream _getExistingFiles: parsed test.log.1 as undefined +0ms
  streamroller:RollingFileWriteStream numToKeep = 4, existingFiles = 1 +1ms
  streamroller:RollingFileWriteStream existing files are:  [ { fileName: 'test.log', index: 0, isCompressed: false } ] +0ms
  streamroller:RollingFileWriteStream writing chunk. file=logs/test.log state={"currentDate":"2019-07-04T18:39:33.490Z","currentIndex":1,"currentSize":0} chunk=[2019-07-04T19:39:40.258] [INFO] default - hello world 1
  streamroller:RollingFileWriteStream  +0ms
```

Looks like it's not detecting `test.log.1` as an existing file

**Comment by nomiddlename** [comment]
Yep, this one's definitely a bug in streamroller. It works fine if you don't restart the app, but you're right it's not picking up old rolled files when it starts up. It'll take me a day or two to fix (I've got only got an hour or two a day when I can work on this).

**Comment by ErisDS** [comment]
I did take a stab at fixing it, but think it needs a little more context than I have / can get from the code. Realise I should probably have put this direct in steamroller - I wasn't sure sorry.

**Comment by nomiddlename** [comment]
The streamroller code is over complicated, there's no shame in not being able to work out what goes on in there. I've got a failing test that reproduces the issue, so I'll get to work on fixing it.

**Comment by ErisDS** [comment]
All good - I just wanted to be clear that I respect your time and at least had a go 😉

**Comment by nomiddlename** [comment]
Merging the PR auto-closed this issue, re-opening it until I can publish to NPM and get verification of the fix.

**Comment by nomiddlename** [comment]
Published to NPM in log4js@4.5.1 - please give it a try and see if it fixes your problem.

**Comment by nomiddlename** [comment]
I've had confirmation from other users that file rotation works properly now, so I'm going to close this one. Please re-open if you think the problem still exists in the latest log4js (5.3.0)
